p4/gateway/gateway.p4

```diff
@@ -53,46 +53,64 @@ control TpsGwIngress(inout headers hdr,
     }
 
     action data_inspect_packet_ipv4(bit<32> device, bit<32> switch_id) {
+        hdr.udp_int.setValid();
         hdr.int_shim.setValid();
         hdr.int_header.setValid();
         hdr.int_meta.setValid();
 
-        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
+        hdr.udp_int.dst_port = UDP_INT_DST_PORT;
+
         hdr.int_shim.next_proto = hdr.ipv4.protocol;
-        hdr.int_shim.type = 1;
-        hdr.int_shim.length = 7;
+        hdr.int_shim.npt = INT_SHIM_NPT_UDP_FULL_WRAP;
+        hdr.int_shim.type = INT_SHIM_TYPE;
+        hdr.int_shim.length = INT_SHIM_BASE_SIZE;
 
-        hdr.int_header.meta_len = 3;
-        hdr.int_header.instructions = 0x80c0;
-        hdr.int_header.remaining_hop_cnt = 10; /* TODO - find a better means to determine this value */
-        hdr.int_header.remaining_hop_cnt = hdr.int_header.remaining_hop_cnt -1;
+        hdr.int_header.ver = INT_VERSION;
+        hdr.int_header.domain_id = INT_SHIM_DOMAIN_ID;
+        hdr.int_header.meta_len = INT_META_LEN;
+        hdr.int_header.instr_bit_0 = TRUE;
+        hdr.int_header.ds_instr_0 = TRUE;
+        hdr.int_header.ds_flags_1 = TRUE;
+        hdr.int_header.remaining_hop_cnt = MAX_HOPS;
 
         hdr.int_meta.switch_id = switch_id;
         hdr.int_meta.orig_mac = hdr.ethernet.src_mac;
 
-        hdr.ipv4.protocol = TYPE_INSPECTION;
-        hdr.ipv4.totalLen = hdr.ipv4.totalLen + ((bit<16>)hdr.int_shim.length * 4);
+        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
+        hdr.ipv4.protocol = TYPE_UDP;
+        hdr.ipv4.totalLen = hdr.ipv4.totalLen + ((bit<16>)hdr.int_shim.length * BYTES_PER_SHIM * INT_SHIM_HOP_SIZE) + UDP_HDR_BYTES;
+
         forwardedPackets.count(device);
     }
 
     action data_inspect_packet_ipv6(bit<32> device, bit<32> switch_id) {
+        hdr.udp_int.setValid();
```

Comment:
Could we check that this packet doesn't already have a UDP Shim header before we add one?

Reply:
The current logic is that gateways are not looking for or parsing INT headers. If that is something we need to support, please create a ticket.